File: scripts/make-lb.py
#!/usr/bin/env python
# make_taaluma.py  –  build A4/A5 PDF from Markdown + template.tex
import sys, os, tempfile, subprocess, shutil, re, helper_functions
from pathlib import Path

def split_title(h1: str):
    """return (part, subtitle) split at first dash"""
    if " - " in h1:
        part, sub = h1.split(" - ", 1)
    elif "-" in h1:
        part, sub = h1.split("-", 1)
    else:
        part, sub = h1, ""
    return part.strip(), sub.strip()

def main():
    md_path = Path("article.md")
    tpl_path = Path(helper_functions.get_git_root() + "/lehachnis-balev/templates/template.tex")
    if not md_path.exists(): sys.exit(f"Markdown '{md_path}' not found")
    if not tpl_path.exists(): sys.exit(f"Template '{tpl_path}' not found")

    # --- read markdown and extract first heading ---------------------------
    with md_path.open(encoding="utf-8") as f:
        lines = f.readlines()

    h1_line = next((ln for ln in lines if ln.lstrip().startswith("#")), "")
    if not h1_line:
        sys.exit("No H1 heading found in markdown")
    h1_text = h1_line.lstrip("#").strip()
    part, subtitle = split_title(h1_text)

    # remove the H1 from markdown body
    body_lines = [ln for ln in lines if ln is not h1_line]

    # --- create temp files --------------------------------------------------
    tmp_dir = tempfile.mkdtemp(prefix="taaluma_")
    md_tmp  = Path(tmp_dir) / "body.md"
    tpl_tmp = Path(tmp_dir) / "template.tex"

    md_tmp.write_text("".join(body_lines), encoding="utf-8")

    # inject title block after \begin{document}
    title_block = rf"""
\begin{{center}}
  {{\headingfont\fontsize{{47}}{{19}}\selectfont {part}\par}}
  {{\headingfont\fontsize{{20}}{{19}}\selectfont {subtitle}\par}}
\end{{center}}\vspace{{1cm}}
"""
    with tpl_path.open(encoding="utf-8") as f_in, tpl_tmp.open("w", encoding="utf-8") as f_out:
        for line in f_in:
            f_out.write(line)
            if line.strip() == r"% <subtitle>":
                f_out.write(title_block)

    # --- run pandoc ---------------------------------------------------------
    pdf_out = md_path.with_suffix(".pdf")
    cmd = [
        "pandoc", str(md_tmp),
        "--template", str(tpl_tmp),
        "--pdf-engine", "xelatex",
        "-o", str(pdf_out)
    ]
    subprocess.check_call(cmd)

    # --- open the PDF -------------------------------------------------------
    print(f"Created {pdf_out}")
    if sys.platform.startswith("win"):
        os.startfile(pdf_out)
    elif sys.platform == "darwin":
        subprocess.Popen(["open", pdf_out])
    else:
        subprocess.Popen(["xdg-open", pdf_out])

if __name__ == "__main__":
    main()

File: scripts/make-lb.rb
#!/usr/bin/env ruby
# make_lb.rb – build A4/A5 PDF from Markdown + template.tex

require 'fileutils'
require 'tmpdir'
require 'pathname'
require 'open3'
require 'debug'
require_relative 'helpers'

# If you have a separate file for this, you can require_relative 'helper_functions'
module HelperFunctions
  def self.get_git_root
    `git rev-parse --show-toplevel`.strip rescue Dir.pwd
  end
end

def main
  if ARGV.length < 2
    abort "usage: make_lb.rb <markdown.md> <template.tex>"
  end

  md_path = Pathname.new(ARGV[0])
  # Building the path similar to your Python version
  tpl_base = Pathname.new(HelperFunctions.get_git_root) + "lehachnis-balev/templates"
  tpl_path = tpl_base + ARGV[1]

  abort "Markdown '#{md_path}' not found" unless md_path.exist?
  abort "Template '#{tpl_path}' not found" unless tpl_path.exist?

  # --- read markdown and extract first heading ---------------------------
  lines = File.readlines(md_path, encoding: "utf-8")

  # --- create temp files --------------------------------------------------
  Dir.mktmpdir("lb_") do |tmp_dir|
    tmp_path = Pathname.new(tmp_dir)
    
    md_tmp   = tmp_path + "body.md"
    tpl_tmp  = tmp_path + "template.tex"

    # Find the first H1 line
    h1_line = lines.find { |ln| ln.lstrip.start_with?("# ") }
    tpl = tpl_path

    if h1_line
      h1_text = h1_line.lstrip.delete_prefix("# ").strip
      part, subtitle = h1_text.split(/\s*-\s*/)

      title_block = <<~LATEX
        \\begin{center}
        {\\headingfont\\fontsize{47}{19}\\selectfont #{part}\\par}
        {\\headingfont\\fontsize{20}{19}\\selectfont #{subtitle}\\par}
        \\end{center}\\vspace{1cm}
      LATEX

      updated_content = File.read(tpl_path, encoding: "utf-8")
        .sub("% <subtitle>", title_block)
      File.write(tpl_tmp, updated_content, encoding: "utf-8")
      tpl = tpl_tmp
    end

    # remove the H1 from markdown body
    body_lines = lines
      .reject { |ln| ln == h1_line }
      .map { _1.gsub(/\(([^\(]*)\)/) { |match| MyHelpers.is_citation($1) ? "`{\\small #{match}}`{=latex}" : match } }

    File.write(md_tmp, body_lines.join, encoding: "utf-8")

    # --- run pandoc ---------------------------------------------------------
    pdf_out = md_path.sub_ext(".pdf")
    # tex_out = md_path.sub_ext(".tex")
    cmd = [
      "pandoc", md_tmp.to_s,
      "--template", tpl.to_s,
      "--pdf-engine", "xelatex",
      "-o", pdf_out.to_s
      # "-o", tex_out.to_s
    ]

    puts "Running Pandoc..."
    puts cmd.join(" ")
    system(*cmd) || abort("Pandoc failed to generate LaTeX file")

    # tex_content = File.read(tex_out, encoding: "utf-8")

    # File.write(tex_out, tex_content, encoding: "utf-8")

    # cmd = [
    #   "xelatex", tex_out.to_s
    # ]

    # system(*cmd) || abort("Pandoc failed to generate PDF")

    # --- open the PDF -------------------------------------------------------
    puts "Created #{pdf_out}"
    case RbConfig::CONFIG['host_os']
    when /mswin|msys|mingw|cygwin|bccwin|wince|emc/
      system("start", pdf_out.to_s)
    when /darwin/
      system("open", pdf_out.to_s)
    else # linux, freebsd, etc.
      system("xdg-open", pdf_out.to_s)
    end
  end
end

if __FILE__ == $0
  main
end
